### DISCUSSION TITLE
Import aliases and Pyright inferred types don't match

### DISCUSSION THREAD

**Discussion by MarkBekooy** [microsoft/pyright#6774] [Q&A]
This might be intended, but I'm not fully certain, so therefore I'd like to ask this question.
It seems like Pyright (Pylance) ignores type aliases when suggesting return types in VS Code.

I just created a random piece of code to provide an example:
![image](https://github.com/microsoft/pyright/assets/24373899/bfe8ede0-4ffd-4d06-bba2-495987f7ae72)

When I double click on this, the `_SupportsCeil` return type is inserted, but that isn't allowed/gets a yellow underline when there is a type alias, as it should be `Ceil`.

So, does Pyright not support these kinds of type aliases? Or is there something else to them?

**Comment by erictraut** [comment]
Yes, this is working as intended.

When you say "type alias", I assume you mean an "import alias" — i.e. an `as` clause in an import statement that assigns the value of an imported symbol to a different local variable name. In your example, the type name is `_SupportsCeil`, but the imported symbol is being stored to the local variable named `Ceil`. The variable `Ceil` is not considered a "type alias" by pyright (or by the Python typing spec). It's simply a local variable that shadows the value of the import. 

Incidentally, you shouldn't be importing private symbols like `_SupportsCeil` from typeshed stubs. They don't exist at runtime, so if you attempt to import them, you'll receive runtime exceptions when you run your code. They're internal implementation details of the stubs, and they are not guaranteed to be stable.

**Reply by MarkBekooy** [reply]
Oops, I'm indeed referring to an "import alias". Nice to know that the imported symbol gets stored in a local variable, but I still find it a bit strange that this local variable isn't then used as a possible type hint.

In my example, as you said, I'm importing a private symbol, which was certainly not the best example I could have given, because those should indeed normally not be imported. But I wanted to show a `from ... import ... as ...` example.

If you have the time, I would like a bit more clarification on the import alias. Let's take a new example:
![image](https://github.com/microsoft/pyright/assets/24373899/7b1402b2-388c-4128-aed6-6d895185fdba)

Here the `defaultdict` is suggested instead of the import alias `DefaultDict`. This is also the case in the generated stub file:
![image](https://github.com/microsoft/pyright/assets/24373899/977b7a76-70b4-4f8a-a6ab-16f5eeefee1d)

Based on this and your previous answer, can I assume that we can ignore import alias local variables (like `DefaultDict`) as type hints? And if we were to actually add a type hint, should it then be `defaultdict` (even though it gets a yellow underline)?

**Reply by erictraut** [reply]
In your example above, the name of the type is `defaultdict`. There may be other ways to refer to the same type, such as import aliases or type aliases, but the original name of the class is `defaultdict`. This is the type that the static type checker tracks, and this is the type that is inferred as the return type of your function. When pylance provides inlay type hints, it's using this inferred type from the type analyzer. 

It's possible that the inferred return type is not imported at all in the current file. In that case, you would need to add an import if you want to insert the type annotation into your code. It's also possible that the type is imported under a different name, in which case you can either import it under its original name or manually adjust the name to refer to an alternate name.

In any case, pyright is working as intended here. It tracks types by their original names even if they are assigned to multiple aliases along the way. It also attempts to preserve type alias names where possible, but it doesn't attempt to track all possible other aliased names to a type.
